packages/core/src/babel-plugin-card-template.ts

```diff
@@ -306,3 +307,84 @@ function updateScope(options: NodePath<t.ObjectExpression>, names: Set<string>,
 
   scope.node.body.properties = scope.node.body.properties.concat(scopeVars);
 }
+
+function buildUsedFieldsListFromUsageMeta(
+  fields: CompiledCard['fields'],
+  defaultFieldFormat: Format,
+  meta: ComponentMeta
+): ComponentInfo['usedFields'] {
+  let usageMeta = meta.usage;
+  let usedFields: Set<string> = new Set();
+
+  if (usageMeta.model && usageMeta.model !== 'self') {
+    for (const fieldPath of usageMeta.model) {
+      usedFields.add(fieldPath);
+    }
+  }
+
+  if (usageMeta.fields === 'self') {
+    usedFields = new Set([...usedFields, ...Object.keys(fields)]);
+  } else {
+    for (const [fieldPath, fieldFormat] of usageMeta.fields.entries()) {
+      buildUsedFieldListFromComponents(
+        usedFields,
+        fieldPath,
+        fields,
+        fieldFormat === 'default' ? defaultFieldFormat : fieldFormat
+      );
+    }
+  }
+
+  return [...usedFields];
+}
+
+function buildUsedFieldListFromComponents(
+  usedFields: Set<string>,
+  fieldPath: string,
+  fields: CompiledCard['fields'],
+  format: Format,
+  prefix?: string
+): void {
+  let field = getFieldForPath(fields, fieldPath);
+
+  if (field && field.card.componentInfos[format].usedFields.length) {
+    for (const nestedFieldPath of field.card.componentInfos[format].usedFields) {
+      buildUsedFieldListFromComponents(usedFields, nestedFieldPath, field.card.fields, 'embedded', fieldPath);
+    }
+  } else {
+    if (prefix) {
+      usedFields.add(`${prefix}.${fieldPath}`);
+    } else {
+      usedFields.add(fieldPath);
+    }
+  }
+}
+
+function canInlineHBS(
+  hasModifiedScope: boolean,
+  usageMeta: TemplateUsageMeta,
```

Comment:
Because of the "default" format of fields, we "need" to use the usageMeta here instead of simply usedFields. This will require us, when we move this logic out of this plugin and into the main part of the compiler to expose the UsageMeta interface. If we want to drop any reference to the field list we could simply use the usedFields list, which would allow us to not have to expose it, but would complicate our ability to specific a format when we invoke a component for a field. Can we update the used fields list to include the format info?

Reply:
I don't think the used field list should gain the format. We don't need to know the answer to canInlineHBS until the transform stage, and by then we should already be able to replace "default" with the real format.